nx/lib/nx/defn/kernel.ex

```diff
@@ -1075,6 +1105,197 @@ defmodule Nx.Defn.Kernel do
+  `Nx.Defn.stream/3`. Let's see an example:
+
+      defmodule Hooks do
+        import Nx.Defn
+
+        defn add_and_mult(a, b) do
+          add = hook(a + b, :hooks_add)
+          mult = hook(a * b, :hooks_mult)
+          {add, mult}
+        end
+      end
+
+  Now you can pass the hook as argument as follows:
+      
+      hooks = %{
+        hooks_add: fn tensor ->
+          IO.inspect {:add, tensor}
+        end
+      }
+
+      args = [Nx.tensor(2), Nx.tensor(3)]
+      Nx.Defn.jit(&Hooks.add_and_mult/2, args, hooks: hooks)
+
+  > **Important!** We recommend to prefix your hook names
+  > by the name of your library to avoid conflicts.
+
+  If a named hook is not given, compilers can optimize
+  that away and not transfer the tensor from the device
+  in the first place.
+
+  You can also mix named hooks with callbacks:
+
+      defn add_and_mult(a, b) do
+        add = hook(a + b, :hooks_add, fn tensor -> IO.inspect({:add, tensor}) end)
+        mult = hook(a * b, :hooks_mult, fn tensor -> IO.inspect({:mult, tensor}) end)
+        {add, mult}
+      end
+
+  If a hook with the same name is given to `Nx.Defn.jit/3`
+  or `Nx.Defn.stream/3`, then it will override the default
+  callback.
+
+  ## Hooks and tokens
+
+  So far, we have always return the result of the `hook`
+  call. However, what happens if the values we want to
+  hook are not part of the return value, such as below?
+
+      defn add_and_mult(a, b) do
+        _add = hook(a + b, :hooks_add, &IO.inspect({:add, &1}))
+        mult = hook(a * b, :hooks_mult, &IO.inspect({:mult, &1}))
+        mult
+      end
+
+  In such cases, you must use tokens. Tokens are used to
+  create an ordering over hooks, ensuring hooks execute
+  in a certain order:
+
+      defn add_and_mult(a, b) do
+        token = create_token()
+        {token, _add} = hook_token(token, a + b, :hooks_add, &IO.inspect({:add, &1}))
+        {token, mult} = hook_token(token, a * b, :hooks_mult, &IO.inspect({:mult, &1}))
+        attach_token(token, mult)
+      end    
+
+  The example above creates a token and uses `hook_token/4`
+  to create hooks attached to said token. By using a token,
+  we guarantee that those hooks will be invoked in the given
+  order. Then, at the end of the function, we attach the token
+  (and the hooks associated with it) to the result `mult`.
+
+  Note you must attach the token at the end, otherwise the hooks
+  will be lost. In fact, the `hook/3` function is implemented
+  roughly like this:
+
+      def hook(tensor_expr, name, function) do
+        {token, result} = hook_token(create_token(), tensor_expr, name, function)
+        attach_token(token, result)
+      end
+
+  """
+  def hook(expr, name, function) when Kernel.and(is_atom(name), is_function(function, 1)),
+    do: unguarded_hook(expr, name, function)
+
+  defp unguarded_hook(expr, name, function) do
+    {token, result} = add_hook(create_token(), expr, name, function)
+    attach_token(token, result)
+  end
```

Comment:
Why do we need to split into this `defp`? Is it used elsewhere?

Reply:
Yes, from hook/2 above.

Reply:
@josevalim I mean, why does it need to be in another function def? Can't it be the body of the guarded `hook/2`?

Reply:
Because we don't want to allow the user to set `function` to nil. It is a value that is only allowed internally (hence the name `unguarded`). :)

Reply:
Oh, I skipped the other 2 definitions above! Nevermind this comment :p